### PR TITLE
Feature/color puzzle

### PR DESCRIPTION
**Pull request by gwenvis**
yo

### CODE REVIEW

File: Assets/Scripts/Puzzles/Color/Player.cs [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+						currentLine = null;
+						currentState = State.Navigating;
+						break;
+					case State.Stuck:
+						running = false;
+						break;
+					case State.Done:
+						running = false;
+						break;
+					default:
```
Comment (stepperman) [inline Assets/Scripts/Puzzles/Color/Player.cs:90]: Codacy Issue found: [Remove this empty 'default' clause.](https://app.codacy.com/manual/stepperman/pvb-programming/pullRequest?prid=5403921)

File: Assets/Scripts/Puzzles/Color/Player.cs [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+		}
+
+		private Line Navigate(ColorCommand colorCommand) => playerPathFinding.FindLine(colorCommand, currentNode);
+
+		private void OnRunCompleted()
+		{
+			Debug.Log($"I AM DONE, ENDED WITH STATE: {currentState}");
+		}
+
+		private IEnumerator RunQueue(Action callback)
```
Comment (stepperman) [inline Assets/Scripts/Puzzles/Color/Player.cs:54]: Codacy Issue found: [This method 'RunQueue' has 42 lines, which is greater than the 30 lines authorized. Split it into smaller methods.](https://app.codacy.com/manual/stepperman/pvb-programming/pullRequest?prid=5403921)

File: Assets/Scripts/Puzzles/Color/UI/ColorCommandDropzone.cs [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+		private struct Item
+		{
+			public DraggableItem draggableItem;
+			public ColorCommand colorCommand;
+			public Transform previousParent;
+		}
+
+		public IEnumerable<ColorCommand> CommandQueue => colors.Select(x=>x.colorCommand);
+
+		private List<Item> colors = new List<Item>();
```
Comment (stepperman) [inline Assets/Scripts/Puzzles/Color/UI/ColorCommandDropzone.cs:22]: Codacy Issue found: [Make 'colors' 'readonly'.](https://app.codacy.com/manual/stepperman/pvb-programming/pullRequest?prid=5403921)

File: Assets/Scripts/Puzzles/Color/Line.cs [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+			Vector3 vector = endNode.transform.position - startingNode.transform.position;
+			float rotation = Mathf.Atan2(
+				vector.y,
+				vector.x) * Mathf.Rad2Deg;
+			rotation = rotation < 0 ? rotation + 360f : rotation;
+			transform.eulerAngles = new Vector3(transform.eulerAngles.x, transform.eulerAngles.y, rotation);
+			return rotation;
+		}
+
+		private float SetLength()
```
Comment (stepperman) [inline Assets/Scripts/Puzzles/Color/Line.cs:56]: Codacy Issue found: [Change return type to 'void'; not a single caller uses the returned value.](https://app.codacy.com/manual/stepperman/pvb-programming/pullRequest?prid=5403921)

File: Assets/Scripts/Puzzles/Color/Player.cs [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+
+				yield return new WaitForEndOfFrame();
+			}
+
+			callback?.Invoke();
+		}
+
+		private IEnumerator Moving(Line line, Node endNode)
+		{
+			;
```
Comment (stepperman) [inline Assets/Scripts/Puzzles/Color/Player.cs:107]: Codacy Issue found: [Remove this empty statement.](https://app.codacy.com/manual/stepperman/pvb-programming/pullRequest?prid=5403921)

File: Assets/Scripts/Puzzles/Color/Player.cs [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+					case State.Idle:
+						currentState = State.Navigating;
+						break;
+					case State.Navigating:
+						colorCommand = commandQueue.RequestNext();
+						currentLine = Navigate(colorCommand);
+						currentState = currentLine == null ? State.Stuck : State.Moving;
+						break;
+					case State.Moving:
+						Node endNode = currentNode == currentLine.StartingNode ? currentLine.EndNode : currentLine.StartingNode;
```
Comment (stepperman) [inline Assets/Scripts/Puzzles/Color/Player.cs:74]: Codacy Issue found: ['currentLine' is null on at least one execution path.](https://app.codacy.com/manual/stepperman/pvb-programming/pullRequest?prid=5403921)

File: Assets/Scripts/Puzzles/Color/ICommandQueue.cs [hunk 7]
```diff
@@ -0,0 +1,8 @@
+﻿namespace DN.Puzzle
+{
+	public interface ICommandQueue<T>
```
Comment (stepperman) [inline Assets/Scripts/Puzzles/Color/ICommandQueue.cs:3]: Codacy Issue found: [Add the 'out' keyword to parameter 'T' to make it 'covariant'.](https://app.codacy.com/manual/stepperman/pvb-programming/pullRequest?prid=5403921)

File: Assets/Scripts/Puzzles/Color/PlayerPathFinding.cs [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+		/// Find the line that is requested by the color command.
+		/// </summary>
+		/// <param name="colorCommand">The color command requested by the player.</param>
+		/// <param name="node">The node that it has to search from</param>
+		/// <returns>A line if it has been found, null if there is no line.</returns>
+		public Line FindLine(ColorCommand colorCommand, Node node)
+		{
+			foreach(Line line in node.ConnectedLines)
+			{
+				if ((line.StartingNode == node || line.CanTraverseBothWays) && line.LineColor == colorCommand.Color)
```
Comment (stepperman) [inline Assets/Scripts/Puzzles/Color/PlayerPathFinding.cs:17]: Codacy Issue found: [Add curly braces around the nested statement(s) in this 'if' block.](https://app.codacy.com/manual/stepperman/pvb-programming/pullRequest?prid=5403921)

Comment (stepperman) [inline Assets/Scripts/Puzzles/Color/PlayerPathFinding.cs:17]: Codacy Issue found: [Refactor this method to add validation of parameter 'colorCommand' before using it.](https://app.codacy.com/manual/stepperman/pvb-programming/pullRequest?prid=5403921)

File: Assets/Scripts/Puzzles/Color/Line.cs (hunk 4 above)
Comment (stepperman) [inline Assets/Scripts/Puzzles/Color/Line.cs:56]: Codacy Issue found: [Change return type to 'void'; not a single caller uses the returned value.](https://app.codacy.com/manual/stepperman/pvb-programming/pullRequest?prid=5403921)

File: Assets/Scripts/Puzzles/Color/Player.cs (hunk 1 above)
Comment (stepperman) [inline Assets/Scripts/Puzzles/Color/Player.cs:90]: Codacy Issue found: [Remove this empty 'default' clause.](https://app.codacy.com/manual/stepperman/pvb-programming/pullRequest?prid=5403921)

File: Assets/Scripts/Puzzles/Color/Player.cs (hunk 6 above)
Comment (stepperman) [inline Assets/Scripts/Puzzles/Color/Player.cs:74]: Codacy Issue found: ['currentLine' is null on at least one execution path.](https://app.codacy.com/manual/stepperman/pvb-programming/pullRequest?prid=5403921)

File: Assets/Scripts/Puzzles/Color/UI/ColorCommandDropzone.cs [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+using UnityEngine;
+
+namespace DN.Puzzle.Color
+{
+	/// <summary>
+	/// ADD CLASS SUMMARY!
+	/// </summary>
+	public class ColorCommandDropzone : MonoBehaviour, IDroppable
+	{
+		private struct Item
```
Comment (stepperman) [inline Assets/Scripts/Puzzles/Color/UI/ColorCommandDropzone.cs:13]: Codacy Issue found: [Implement 'IEquatable<T>' in value type 'Item'.](https://app.codacy.com/manual/stepperman/pvb-programming/pullRequest?prid=5403921)

File: Assets/Scripts/Puzzles/Color/PlayerPathFinding.cs (hunk 8 above)
Comment (stepperman) [inline Assets/Scripts/Puzzles/Color/PlayerPathFinding.cs:17]: Codacy Issue found: [Add curly braces around the nested statement(s) in this 'if' block.](https://app.codacy.com/manual/stepperman/pvb-programming/pullRequest?prid=5403921)

File: Assets/Scripts/Puzzles/Color/ColorPuzzleSettings.cs [hunk 10]
```diff
@@ -0,0 +1,30 @@
+﻿using UnityEngine;
+
+namespace DN.Puzzle.Color
+{
+	public class ColorPuzzleSettings : ScriptableObject
+	{
+		[System.Serializable]
+		public struct ColorLineSettings
```
Comment (stepperman) [inline Assets/Scripts/Puzzles/Color/ColorPuzzleSettings.cs:8]: Codacy Issue found: [Implement 'IEquatable<T>' in value type 'ColorLineSettings'.](https://app.codacy.com/manual/stepperman/pvb-programming/pullRequest?prid=5403921)

File: Assets/Scripts/Puzzles/Color/PlayerPathFinding.cs (hunk 8 above)
Comment (stepperman) [inline Assets/Scripts/Puzzles/Color/PlayerPathFinding.cs:17]: Codacy Issue found: [Refactor this method to add validation of parameter 'colorCommand' before using it.](https://app.codacy.com/manual/stepperman/pvb-programming/pullRequest?prid=5403921)

File: Assets/Scripts/Puzzles/Color/PlayerPathFinding.cs [hunk 11]
```diff
@@ -0,0 +1,24 @@
...
+{
+	public class PlayerPathFinding : MonoBehaviour
+	{ 
+		/// <summary>
+		/// Find the line that is requested by the color command.
+		/// </summary>
+		/// <param name="colorCommand">The color command requested by the player.</param>
+		/// <param name="node">The node that it has to search from</param>
+		/// <returns>A line if it has been found, null if there is no line.</returns>
+		public Line FindLine(ColorCommand colorCommand, Node node)
```
Comment (stepperman) [inline Assets/Scripts/Puzzles/Color/PlayerPathFinding.cs:13]: Codacy Issue found: [Make 'FindLine' a static method.](https://app.codacy.com/manual/stepperman/pvb-programming/pullRequest?prid=5403921)

File: Assets/Scripts/Puzzles/Color/Line.cs [hunk 12]
```diff
@@ -0,0 +1,104 @@
...
+
+			Vector3 scale = lineSprite.transform.localScale;
+			scale.x = Vector2.Distance(startingNode.transform.position, endNode.transform.position) / canvas.transform.lossyScale.x;
+			lineSprite.transform.localScale = scale;
+			return scale.x;
+		}
+
+		private void CreateArrow(float rotation, UnityEngine.Color color)
+		{
+			if(CanTraverseBothWays)
```
Comment (stepperman) [inline Assets/Scripts/Puzzles/Color/Line.cs:68]: Codacy Issue found: [Add curly braces around the nested statement(s) in this 'if' block.](https://app.codacy.com/manual/stepperman/pvb-programming/pullRequest?prid=5403921)

File: Assets/Scripts/Puzzles/Color/ICommandQueue.cs (hunk 7 above)
Comment (stepperman) [inline Assets/Scripts/Puzzles/Color/ICommandQueue.cs:3]: Codacy Issue found: [Add the 'out' keyword to parameter 'T' to make it 'covariant'.](https://app.codacy.com/manual/stepperman/pvb-programming/pullRequest?prid=5403921)

File: Assets/Scripts/Puzzles/Color/PlayerPathFinding.cs [hunk 13]
```diff
@@ -0,0 +1,24 @@
...
+		{
+			foreach(Line line in node.ConnectedLines)
```
Comment (stepperman) [inline Assets/Scripts/Puzzles/Color/PlayerPathFinding.cs:15]: Codacy Issue found: [Refactor this method to add validation of parameter 'node' before using it.](https://app.codacy.com/manual/stepperman/pvb-programming/pullRequest?prid=5403921)

File: Assets/Scripts/Puzzles/Color/ColorCommandQueue.cs [hunk 14]
```diff
@@ -0,0 +1,19 @@
+﻿using System.Collections.Generic;
+using System.Linq;
+
+namespace DN.Puzzle.Color
+{
+	public class ColorCommandQueue : ICommandQueue<ColorCommand>
+	{
+		private Queue<ColorCommand> colorCommands;
```
Comment (stepperman) [inline Assets/Scripts/Puzzles/Color/ColorCommandQueue.cs:8]: Codacy Issue found: [Make 'colorCommands' 'readonly'.](https://app.codacy.com/manual/stepperman/pvb-programming/pullRequest?prid=5403921)

File: Assets/Scripts/Puzzles/Color/Player.cs (hunk 5 above)
Comment (stepperman) [inline Assets/Scripts/Puzzles/Color/Player.cs:107]: Codacy Issue found: [Remove this empty statement.](https://app.codacy.com/manual/stepperman/pvb-programming/pullRequest?prid=5403921)

File: Assets/Scripts/Puzzles/Color/UI/ColorCommandDropzone.cs (hunk 3 above)
Comment (stepperman) [inline Assets/Scripts/Puzzles/Color/UI/ColorCommandDropzone.cs:22]: Codacy Issue found: [Make 'colors' 'readonly'.](https://app.codacy.com/manual/stepperman/pvb-programming/pullRequest?prid=5403921)

File: Assets/Scripts/Puzzles/Color/UI/ColorCommandDropzone.cs [hunk 15]
```diff
@@ -0,0 +1,45 @@
...
+		public IEnumerable<ColorCommand> CommandQueue => colors.Select(x=>x.colorCommand);
+
+		private List<Item> colors = new List<Item>();
+
+		public void Drop(DraggableItem droppedObject)
+		{
+			colors.Add(new Item
+			{
+				draggableItem = droppedObject,
+				previousParent = droppedObject.transform.parent,
```
Comment (stepperman) [inline Assets/Scripts/Puzzles/Color/UI/ColorCommandDropzone.cs:29]: Codacy Issue found: [Refactor this method to add validation of parameter 'droppedObject' before using it.](https://app.codacy.com/manual/stepperman/pvb-programming/pullRequest?prid=5403921)

File: Assets/Scripts/Puzzles/Color/Player.cs (hunk 2 above)
Comment (stepperman) [inline Assets/Scripts/Puzzles/Color/Player.cs:54]: Codacy Issue found: [This method 'RunQueue' has 42 lines, which is greater than the 30 lines authorized. Split it into smaller methods.](https://app.codacy.com/manual/stepperman/pvb-programming/pullRequest?prid=5403921)

File: Assets/Scripts/Puzzles/Color/Node.cs [hunk 16]
```diff
@@ -0,0 +1,30 @@
+﻿using System.Collections.Generic;
+using UnityEngine;
+
+namespace DN.Puzzle.Color
+{
+	/// <summary>
+	/// 
+	/// </summary>
```
Comment (MeesMD) [inline Assets/Scripts/Puzzles/Color/Node.cs:8]: empty summary? sciance man does not approve >:(

File: Assets/Scripts/Puzzles/Color/ColorCommand.cs [hunk 17]
```diff
@@ -0,0 +1,11 @@
+﻿using UnityEngine;
+
+namespace DN.Puzzle.Color
+{
+	public class ColorCommand : MonoBehaviour
```
Comment (MeesMD) [inline Assets/Scripts/Puzzles/Color/ColorCommand.cs:5]: Where summary?

File: Assets/Scripts/Puzzles/Color/UI/ColorCommandDropzone.cs [hunk 18]
```diff
@@ -0,0 +1,45 @@
+﻿using DN.UI;
+using System.Collections.Generic;
+using System.Linq;
+using UnityEngine;
+
+namespace DN.Puzzle.Color
+{
+	/// <summary>
+	/// ADD CLASS SUMMARY!
+	/// </summary>
```
Comment (MeesMD) [inline Assets/Scripts/Puzzles/Color/UI/ColorCommandDropzone.cs:10]: ??

File: Assets/Scripts/Puzzles/Color/UI/RunButton.cs [hunk 19]
```diff
@@ -0,0 +1,30 @@
+﻿using UnityEngine;
+
+namespace DN.Puzzle.Color.UI
+{
+	/// <summary>
+	/// ADD CLASS SUMMARY!
+	/// </summary>
```
Comment (MeesMD) [inline Assets/Scripts/Puzzles/Color/UI/RunButton.cs]: ??

File: Assets/Scripts/Puzzles/Color/ColorCommandQueue.cs [hunk 20]
```diff
@@ -0,0 +1,19 @@
+﻿using System.Collections.Generic;
+using System.Linq;
+
+namespace DN.Puzzle.Color
+{
+	public class ColorCommandQueue : ICommandQueue<ColorCommand>
```
Comment (MeesMD) [inline Assets/Scripts/Puzzles/Color/ColorCommandQueue.cs]: where summary?

File: Assets/Scripts/Puzzles/Color/ColorPuzzleSettings.cs [hunk 21]
```diff
@@ -0,0 +1,30 @@
+﻿using UnityEngine;
+
+namespace DN.Puzzle.Color
+{
+	public class ColorPuzzleSettings : ScriptableObject
```
Comment (MeesMD) [inline Assets/Scripts/Puzzles/Color/ColorPuzzleSettings.cs]: summary?

File: Assets/Scripts/Puzzles/Color/Line.cs [hunk 22]
```diff
@@ -0,0 +1,104 @@
+﻿using System.Linq;
+using UnityEngine;
+using UnityEngine.UI;
+
+namespace DN.Puzzle.Color
+{
+	public class Line : MonoBehaviour
+	{
```
Comment (MeesMD) [inline Assets/Scripts/Puzzles/Color/Line.cs]: Doing summary pls

**Comment by stepperman** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
- Added 10
           

Complexity increasing per file
==============================
- Assets/Scripts/Puzzles/Color/Node.cs  1
- Assets/Scripts/Puzzles/Color/ColorCommand.cs  1
- Assets/Scripts/Puzzles/Color/UI/ColorCommandDropzone.cs  2
- Assets/Scripts/Puzzles/Color/ColorPuzzleSettings.cs  1
- Assets/Scripts/Puzzles/Color/UI/RunButton.cs  1
- Assets/Scripts/Puzzles/Color/Player.cs  14
- Assets/Scripts/Puzzles/Color/ColorCommandQueue.cs  1
- Assets/Scripts/Puzzles/Color/PlayerPathFinding.cs  5
- Assets/Scripts/Puzzles/Color/Line.cs  4
- Assets/Scripts/DraggableItem.cs  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/stepperman/pvb-programming/pullRequest?prid=5403921&bid=17714615)

**Comment by stepperman** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
- Added 9
           

Complexity increasing per file
==============================
- Assets/Scripts/Puzzles/Color/ColorPuzzleSettings.cs  1
- Assets/Scripts/DraggableItem.cs  1
- Assets/Scripts/Puzzles/Color/ColorCommandQueue.cs  1
- Assets/Scripts/Puzzles/Color/PlayerPathFinding.cs  5
- Assets/Scripts/Puzzles/Color/Line.cs  4
- Assets/Scripts/Puzzles/Color/ColorCommand.cs  1
- Assets/Scripts/Puzzles/Color/UI/ColorCommandDropzone.cs  2
- Assets/Scripts/Puzzles/Color/Player.cs  14
- Assets/Scripts/Puzzles/Color/Node.cs  1
- Assets/Scripts/Puzzles/Color/UI/RunButton.cs  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/stepperman/pvb-programming/pullRequest?prid=5403921&bid=17714615)

**Comment by stepperman** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
- Added 9
           

Complexity increasing per file
==============================
- Assets/Scripts/Puzzles/Color/ColorPuzzleSettings.cs  1
- Assets/Scripts/Puzzles/Color/UI/RunButton.cs  1
- Assets/Scripts/Puzzles/Color/Player.cs  14
- Assets/Scripts/Puzzles/Color/PlayerPathFinding.cs  5
- Assets/Scripts/Puzzles/Color/ColorCommandQueue.cs  1
- Assets/Scripts/DraggableItem.cs  1
- Assets/Scripts/Puzzles/Color/Node.cs  1
- Assets/Scripts/Puzzles/Color/Line.cs  4
- Assets/Scripts/Puzzles/Color/ColorCommand.cs  1
- Assets/Scripts/Puzzles/Color/UI/ColorCommandDropzone.cs  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/stepperman/pvb-programming/pullRequest?prid=5403921&bid=17714615)

**Comment by stepperman** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
- Added 9
           

Complexity increasing per file
==============================
- Assets/Scripts/DraggableItem.cs  1
- Assets/Scripts/Puzzles/Color/Player.cs  14
- Assets/Scripts/Puzzles/Color/PlayerPathFinding.cs  5
- Assets/Scripts/Puzzles/Color/UI/ColorCommandDropzone.cs  2
- Assets/Scripts/Puzzles/Color/Line.cs  4
- Assets/Scripts/Puzzles/Color/ColorCommandQueue.cs  1
- Assets/Scripts/Puzzles/Color/UI/FinishUI.cs  4
- Assets/Scripts/Puzzles/Color/ColorPuzzleSettings.cs  1
- Assets/Scripts/Puzzles/Color/UI/RunButton.cs  1
- Assets/Scripts/Puzzles/Color/ColorCommand.cs  1
- Assets/Scripts/Puzzles/Color/Node.cs  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/stepperman/pvb-programming/pullRequest?prid=5403921&bid=17714615)

**Comment by stepperman** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
- Added 9
           

Complexity increasing per file
==============================
- Assets/Scripts/DraggableItem.cs  1
- Assets/Scripts/Puzzles/Color/PlayerPathFinding.cs  5
- Assets/Scripts/Puzzles/Color/Line.cs  4
- Assets/Scripts/Puzzles/Color/UI/RunButton.cs  1
- Assets/Scripts/Puzzles/Color/ColorCommandQueue.cs  1
- Assets/Scripts/Puzzles/Color/UI/FinishUI.cs  4
- Assets/Scripts/Puzzles/Color/Player.cs  14
- Assets/Scripts/Puzzles/Color/UI/ColorCommandDropzone.cs  2
- Assets/Scripts/Puzzles/Color/ColorCommand.cs  1
- Assets/Scripts/Puzzles/Color/Node.cs  1
- Assets/Scripts/Puzzles/Color/ColorPuzzleSettings.cs  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/stepperman/pvb-programming/pullRequest?prid=5403921&bid=17714615)

**Review by joppevanbasten — APPROVED** [review]

**Pull request closed without merging**